Add initial instrumentation of kafka connect SinkTask

vasantteja opened the pull request:
## 📋 **PR Description**

### **Overview**
This PR adds OpenTelemetry instrumentation for Apache Kafka Connect SinkTask operations, providing observability into data pipeline processing from Kafka topics to external systems.

### **What This Adds**
- **Instrumentation for `SinkTask.put()` method**: Creates spans for batch processing operations
- **Trace context propagation**: Links producer spans to sink processing spans using Kafka message headers
- **Comprehensive integration tests**: Docker-based tests with MongoDB and PostgreSQL sinks
- **Support for Kafka Connect 2.6+**: Compatible with modern Kafka Connect deployments

### **Key Features**

#### 🔍 **Span Creation**
- **Span name**: `KafkaConnect.put`
- **Span kind**: `INTERNAL`
- **Attributes**: Thread information (`thread.name`, `thread.id`)
- **Links**: Creates span links to producer spans extracted from Kafka message headers

#### 🔗 **Trace Context Propagation**
- Extracts trace context from Kafka message headers using `TextMapPropagator`
- Links sink processing spans to original producer spans for end-to-end tracing
- Handles batch processing scenarios with multiple records

#### 🧪 **Testing Approach**
Uses **service-side instrumentation testing pattern** (similar to JMX metrics instrumentation):
- **Docker containers** for Kafka Connect, Kafka, and sink databases
- **OTLP backend** to collect spans from containerized services
- **Integration tests** with real MongoDB and PostgreSQL connectors
- **Robust container lifecycle management** with proper cleanup

### **Technical Implementation**

#### **Instrumentation Details**
```java
// Instruments all SinkTask implementations
@Override
public ElementMatcher<TypeDescription> typeMatcher() {
  return hasSuperType(named("org.apache.kafka.connect.sink.SinkTask"));
}

// Instruments the put() method
@Override
public void transform(TypeTransformer transformer) {
  transformer.applyAdviceToMethod(
    isPublic().and(named("put")).and(takesArgument(0, Collection.class)),
    SinkTaskInstrumentation.class.getName() + "$SinkTaskPutAdvice");
}
```

#### **Context Extraction**
- Uses `PropagatorBasedSpanLinksExtractor` for creating span links
- Implements `SinkRecordHeadersGetter` for extracting headers from Kafka records
- Handles classloader isolation challenges in Kafka Connect plugin architecture

#### **Classloader Considerations**
Kafka Connect's `PluginClassLoader` isolation required the **Singletons Static Reference pattern**:
```java
// Access TextMapGetter through singletons to avoid classloader issues
Context extractedContext = KafkaConnectSingletons.propagator()
    .extract(parentContext, firstRecord, KafkaConnectSingletons.sinkRecordHeaderGetter());
```

### **Testing Strategy**

#### **Why Not `InstrumentationExtension`?**
Unlike client-side instrumentations, Kafka Connect runs as a **separate service** in its own JVM. The standard `InstrumentationExtension` cannot collect spans from containerized services, requiring an **OTLP-based approach**.

#### **Test Architecture**
```
┌─────────────────┐    ┌──────────────┐    ┌─────────────────┐
│   Test JVM      │    │ Kafka Connect│    │  OTLP Backend   │
│                 │    │  Container    │    │   Container     │
│ • Produces msgs │───▶│ • Processes   │───▶│ • Collects      │
│ • Verifies spans│    │ • Instruments │    │   spans         │
└─────────────────┘    └──────────────┘    └─────────────────┘
```

#### **Test Coverage**
- ✅ **MongoDB Sink**: Tests with MongoDB connector and document insertion
- ✅ **PostgreSQL Sink**: Tests with JDBC connector and table insertion  
- ✅ **Span verification**: Validates span names, attributes, and trace linking
- ✅ **Container lifecycle**: Proper setup, execution, and cleanup
- ✅ **ARM64 compatibility**: Handles Docker architecture emulation with increased timeouts

### **Performance Considerations**
- **Minimal overhead**: Only instruments the `put()` method during batch processing
- **Efficient context extraction**: Extracts context only from the first record in batch
- **Proper resource cleanup**: All containers and connections are properly closed

### **Future Enhancements**
- Support for source tasks (`SourceTask` instrumentation)

### **Testing Instructions**

```bash
# Run all Kafka Connect tests
./gradlew :instrumentation:kafka:kafka-connect-2.6:testing:test

# Run specific tests
./gradlew :instrumentation:kafka:kafka-connect-2.6:testing:test --tests="*MongoKafkaConnectSinkTaskTest*"
./gradlew :instrumentation:kafka:kafka-connect-2.6:testing:test --tests="*PostgresKafkaConnectSinkTaskTest*"

# Build the instrumentation
./gradlew :instrumentation:kafka:kafka-connect-2.6:javaagent:build
```

### **Related Issues**
1. https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/12322
2. https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/12261

Thanks in advance and please let me know your thoughts.

breedx-splk commented on settings.gradle.kts:
```diff
@@ -361,7 +362,12 @@ include(":instrumentation:kafka:kafka-clients:kafka-clients-0.11:javaagent")
 include(":instrumentation:kafka:kafka-clients:kafka-clients-0.11:testing")
 include(":instrumentation:kafka:kafka-clients:kafka-clients-2.6:library")
 include(":instrumentation:kafka:kafka-clients:kafka-clients-common-0.11:library")
+include(":instrumentation:kafka:kafka-clients:kafka-clients-common-0.11:testing")
+include(":instrumentation:kafka:kafka-connect-2.6:javaagent")
+include(":instrumentation:kafka:kafka-connect-2.6:testing")
+include(":instrumentation:kafka:kafka-streams-0.11:bootstrap")
 include(":instrumentation:kafka:kafka-streams-0.11:javaagent")
+include(":instrumentation:kotlinx-coroutines:javaagent")
```
This PR seems to only add kafka-connect-2.6, so I'm surprised to see these changes include all these references in the topmost settings.gradle.kts. Any idea why or what I'm missing?

jaydeluca commented on instrumentation/kafka/kafka-connect-2.6/javaagent/README.md:
```diff
@@ -0,0 +1,101 @@
+# Kafka Connect SinkTask Integration Test
```
We don't typically add readme's for test classes. A readme that describes the instrumentation would be more useful, if you want to take a look at how some of the other instrumentations are documented

jaydeluca commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/KafkaConnectInstrumentationModule.java:
```diff
@@ -0,0 +1,32 @@
+
+  public KafkaConnectInstrumentationModule() {
+    super("kafka-connect", "kafka-connect-2.6");
+    logger.info("KafkaConnect: InstrumentationModule constructor called");
```
can you remove the logging statements from this class?

jaydeluca commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/SinkTaskInstrumentation.java:
```diff
@@ -0,0 +1,113 @@
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    logger.info("KafkaConnect: typeMatcher called");
```
can you remove the logging from this class too?

jaydeluca commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/SinkTaskInstrumentation.java:
```diff
@@ -0,0 +1,113 @@
+
+    // Add advice to constructor to see what classes we're instrumenting
+    transformer.applyAdviceToMethod(
+        named("<init>"), SinkTaskInstrumentation.class.getName() + "$ConstructorAdvice");
```
this and the `ConstructorAdvice` look like they might potentially be leftover debugging code that should be removed?

vasantteja replied:
Removed this!!

jaydeluca commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,42 @@
+  testImplementation("org.testcontainers:kafka:1.19.7")
+  implementation("io.rest-assured:rest-assured:5.5.5")
+  testImplementation("org.junit.jupiter:junit-jupiter:5.10.2")
+  // Remove this duplicate line: testImplementation("org.apache.httpcomponents:httpclient:4.3")
```
should this comment be removed?

```suggestion
```

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/build.gradle.kts:
```diff
@@ -0,0 +1,40 @@
+    systemProperty("junit.jupiter.execution.timeout.default", "5m")
+  }
+
+  withType<JavaCompile>().configureEach {
+    options.compilerArgs.add("-Xlint:-deprecation")
+  }
```
usually we use `@SupressWarnings` where needed

laurit commented on settings.gradle.kts:
```diff
@@ -102,6 +102,7 @@ include(":test-report")
 include(":testing:agent-exporter")
 include(":testing:agent-for-testing")
 include(":testing:armeria-shaded-for-testing")
+
```
remove this line

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/build.gradle.kts:
```diff
@@ -0,0 +1,40 @@
+
+    // Enable experimental span attributes and receive telemetry for comprehensive testing
+    jvmArgs("-Dotel.instrumentation.kafka.experimental-span-attributes=true")
+    jvmArgs("-Dotel.instrumentation.messaging.experimental.receive-telemetry.enabled=true")
```
just curious, does using this flag change anything for the tests you have?

vasantteja replied:
This doesn't change anything. I added it when debugging the stuff and kept it. I removed it and tests ran fine.

laurit reviewed (commented):
Do I understand correctly that this instrumentation only creates a span that is linked to the producer when the message is consumed by the sink. Or does it do something else like allow for the trace to propagate to where the message is read from where the sink sent it? I don't know anything about kafka connect so sorry if this didn't make sense.

vasantteja commented:
> Do I understand correctly that this instrumentation only creates a span that is linked to the producer when the message is consumed by the sink. Or does it do something else like allow for the trace to propagate to where the message is read from where the sink sent it? I don't know anything about kafka connect so sorry if this didn't make sense.

@laurit Great question!! Yes, this instrumentation creates a span that is linked to the producer when the message is consumed by the sink. It also allows for the trace to propagate to downstream operations.

### How it works:
Span Linking: Creates links back to all producer spans in the batch for correlation
Trace Propagation: Makes the trace context current so downstream operations inherit it

### Examples:
JDBC Kafka Connector: [Uses](https://github.com/confluentinc/kafka-connect-jdbc/blob/20c9998473e8c181115f9ebd2f7ca93600400c59/src/main/java/io/confluent/connect/jdbc/sink/BufferedRecords.java#L198) `PreparedStatement.executeBatch()`. Since JDBC is instrumented, we create a parent-child relationship between the Kafka Connect span and the executeBatch() span.
MongoDB Kafka Connector: [Uses](https://github.com/mongodb/mongo-kafka/blob/ba92d40f327334ebb55e9034010b8c852a318d85/src/main/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTask.java#L162) `collection.bulkWrite()`. MongoDB's command-level instrumentation creates parent-child relationships between the Kafka Connect span and the resulting MongoDB command spans (insert/update/delete).

### Result: 
We get both span linking (for correlation back to producers) and trace propagation (for continuity to databases) - enabling end-to-end tracing through the entire data pipeline.

Let me know if this answers your question!!

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/javaagent/build.gradle.kts:
```diff
@@ -0,0 +1,40 @@
+    jvmArgs("-Dotel.instrumentation.kafka.experimental-span-attributes=true")
+    jvmArgs("-Dotel.instrumentation.messaging.experimental.receive-telemetry.enabled=true")
+    // Set timeout for integration tests with containers
+    systemProperty("junit.jupiter.execution.timeout.default", "5m")
```
This doesn't provide any additional value. I added it thinking there is no global duration. Thanks for this.

vasantteja commented:
> > @laurit Great question!! Yes, this instrumentation creates a span that is linked to the producer when the message is consumed by the sink. However, it depends on the specific connector implementation when it comes to automatic trace propagation to downstream databases. If we have instrumentation for the database operation that the connector uses, automatic trace propagation happens; if not, the trace is not propagated beyond the Kafka Connect span.
> 
> Generally we don't expect databases to propagate context. Even when the context is propagated to the database the intention is to tag the database query. For example context propagation to sql database could be used to get the trace id for a slow query so you could track down what executed that query.
> 
> > ### Examples:
> > **JDBC Kafka Connector**: [Uses](https://github.com/confluentinc/kafka-connect-jdbc/blob/20c9998473e8c181115f9ebd2f7ca93600400c59/src/main/java/io/confluent/connect/jdbc/sink/BufferedRecords.java#L198) `PreparedStatement.executeBatch()`. Since `PreparedStatement.executeBatch()` is [explicitly excluded](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jdbc/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jdbc/PreparedStatementInstrumentation.java#L58) from JDBC instrumentation, we don't see a parent-child relationship between the Kafka Connect span and the database operation.
> 
> You are misreading this. https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jdbc/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jdbc/PreparedStatementInstrumentation.java#L58 excludes these methods because they are instrumented by
> 
> https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/eb7ea9ed538d31e28cac2f93ed2af2870063e3ac/instrumentation/jdbc/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jdbc/StatementInstrumentation.java#L57
> 
> > **Summary**: This instrumentation provides span linking (connecting to producer spans) and sets up the foundation for trace propagation by making the Kafka Connect span the active context. Whether downstream spans are created depends entirely on whether the specific database operations used by each connector are instrumented by OpenTelemetry.
> 
> I'm wondering why you chose to create an internal span, wouldn't a consumer span be more appropriate?

@laurit Thanks for the reply!! Appreciate the help and thanks for correcting me.

There are three things which I want to breakdown here:
1. Thanks for the context on the database part. It makes sense now. So I believe if we link the producer trace with kafka-connect we are good.
2. I am sorry I was mistaken on the JDBC instrumentation. In fact, I wrote a test to confirm if it's being instrumented or not and it looks like it's being instrumented. Here are the logs from the container showing the complete trace flow:
```
[otel.javaagent 2025-09-02 22:14:56:041 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'SELECT test' : faa1a2f506b412fc46b6daf1c8bb9bce 36869b3ce35d42f3 CLIENT [tracer: io.opentelemetry.jdbc:2.20.0-alpha-SNAPSHOT] AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT NULL AS TABLE_CAT, n.nspname AS TABLE_SCHEM, c.relname AS TABLE_NAME, CASE n.nspname ~ ? OR n.nspname = ? WHEN true THEN CASE WHEN n.nspname = ? OR n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END WHEN n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END END WHEN false THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? then ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE NULL END AS TABLE_TYPE, d.description AS REMARKS, ? as TYPE_CAT, ? as TYPE_SCHEM, ? as TYPE_NAME, ? AS SELF_REFERENCING_COL_NAME, ? AS REF_GENERATION FROM pg_catalog.pg_namespace n, pg_catalog.pg_class c LEFT JOIN pg_catalog.pg_description d ON (c.oid = d.objoid AND d.objsubid = ? and d.classoid = ?::regclass) WHERE c.relnamespace = n.oid AND c.relname LIKE ? AND (false OR ( c.relkind = ? AND n.nspname !~ ? AND n.nspname <> ? ) ) ORDER BY TABLE_TYPE,TABLE_SCHEM,TABLE_NAME }, capacity=128, totalAddedValues=10}
[otel.javaagent 2025-09-02 22:14:56:041 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.javaagent.testing.exporter.OtlpInMemorySpanExporter - Exporting span SpanData{spanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=36869b3ce35d42f3, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, parentSpanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=ebe3a0102b0469cf, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, resource=Resource{schemaUrl=null, attributes={service.instance.id="5b2f150f-21b3-452e-b15d-5af413ba2da8", service.name="unknown_service:java", telemetry.distro.name="opentelemetry-java-instrumentation", telemetry.distro.version="2.20.0-SNAPSHOT", telemetry.sdk.language="java", telemetry.sdk.name="opentelemetry", telemetry.sdk.version="1.53.0"}}, instrumentationScopeInfo=InstrumentationScopeInfo{name=io.opentelemetry.jdbc, version=2.20.0-alpha-SNAPSHOT, schemaUrl=null, attributes={}}, name=SELECT test, kind=CLIENT, startEpochNanos=1756851296025217709, endEpochNanos=1756851296040659750, attributes=AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT NULL AS TABLE_CAT, n.nspname AS TABLE_SCHEM, c.relname AS TABLE_NAME, CASE n.nspname ~ ? OR n.nspname = ? WHEN true THEN CASE WHEN n.nspname = ? OR n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END WHEN n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END END WHEN false THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? then ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE NULL END AS TABLE_TYPE, d.description AS REMARKS, ? as TYPE_CAT, ? as TYPE_SCHEM, ? as TYPE_NAME, ? AS SELF_REFERENCING_COL_NAME, ? AS REF_GENERATION FROM pg_catalog.pg_namespace n, pg_catalog.pg_class c LEFT JOIN pg_catalog.pg_description d ON (c.oid = d.objoid AND d.objsubid = ? and d.classoid = ?::regclass) WHERE c.relnamespace = n.oid AND c.relname LIKE ? AND (false OR ( c.relkind = ? AND n.nspname !~ ? AND n.nspname <> ? ) ) ORDER BY TABLE_TYPE,TABLE_SCHEM,TABLE_NAME }, capacity=128, totalAddedValues=10}, totalAttributeCount=10, events=[], totalRecordedEvents=0, links=[], totalRecordedLinks=0, status=ImmutableStatusData{statusCode=UNSET, description=}, hasEnded=true}
[2025-09-02 22:14:56,041] INFO Using PostgreSql dialect TABLE "person" present (io.confluent.connect.jdbc.dialect.GenericDatabaseDialect)
[otel.javaagent 2025-09-02 22:14:56:049 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'SELECT test' : faa1a2f506b412fc46b6daf1c8bb9bce 6d767da1a2b935f7 CLIENT [tracer: io.opentelemetry.jdbc:2.20.0-alpha-SNAPSHOT] AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT result.TABLE_CAT, result.TABLE_SCHEM, result.TABLE_NAME, result.COLUMN_NAME, result.KEY_SEQ, result.PK_NAME FROM (SELECT NULL AS TABLE_CAT, n.nspname AS TABLE_SCHEM, ct.relname AS TABLE_NAME, a.attname AS COLUMN_NAME, (information_schema._pg_expandarray(i.indkey)).n AS KEY_SEQ, ci.relname AS PK_NAME, information_schema._pg_expandarray(i.indkey) AS KEYS, a.attnum AS A_ATTNUM FROM pg_catalog.pg_class ct JOIN pg_catalog.pg_attribute a ON (ct.oid = a.attrelid) JOIN pg_catalog.pg_namespace n ON (ct.relnamespace = n.oid) JOIN pg_catalog.pg_index i ON ( a.attrelid = i.indrelid) JOIN pg_catalog.pg_class ci ON (ci.oid = i.indexrelid) WHERE true AND ct.relname = ? AND i.indisprimary ) result where result.A_ATTNUM = (result.KEYS).x ORDER BY result.table_name, result.pk_name, result.key_seq}, capacity=128, totalAddedValues=10}
[otel.javaagent 2025-09-02 22:14:56:050 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.javaagent.testing.exporter.OtlpInMemorySpanExporter - Exporting span SpanData{spanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=6d767da1a2b935f7, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, parentSpanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=ebe3a0102b0469cf, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, resource=Resource{schemaUrl=null, attributes={service.instance.id="5b2f150f-21b3-452e-b15d-5af413ba2da8", service.name="unknown_service:java", telemetry.distro.name="opentelemetry-java-instrumentation", telemetry.distro.version="2.20.0-SNAPSHOT", telemetry.sdk.language="java", telemetry.sdk.name="opentelemetry", telemetry.sdk.version="1.53.0"}}, instrumentationScopeInfo=InstrumentationScopeInfo{name=io.opentelemetry.jdbc, version=2.20.0-alpha-SNAPSHOT, schemaUrl=null, attributes={}}, name=SELECT test, kind=CLIENT, startEpochNanos=1756851296042146375, endEpochNanos=1756851296049224084, attributes=AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT result.TABLE_CAT, result.TABLE_SCHEM, result.TABLE_NAME, result.COLUMN_NAME, result.KEY_SEQ, result.PK_NAME FROM (SELECT NULL AS TABLE_CAT, n.nspname AS TABLE_SCHEM, ct.relname AS TABLE_NAME, a.attname AS COLUMN_NAME, (information_schema._pg_expandarray(i.indkey)).n AS KEY_SEQ, ci.relname AS PK_NAME, information_schema._pg_expandarray(i.indkey) AS KEYS, a.attnum AS A_ATTNUM FROM pg_catalog.pg_class ct JOIN pg_catalog.pg_attribute a ON (ct.oid = a.attrelid) JOIN pg_catalog.pg_namespace n ON (ct.relnamespace = n.oid) JOIN pg_catalog.pg_index i ON ( a.attrelid = i.indrelid) JOIN pg_catalog.pg_class ci ON (ci.oid = i.indexrelid) WHERE true AND ct.relname = ? AND i.indisprimary ) result where result.A_ATTNUM = (result.KEYS).x ORDER BY result.table_name, result.pk_name, result.key_seq}, capacity=128, totalAddedValues=10}, totalAttributeCount=10, events=[], totalRecordedEvents=0, links=[], totalRecordedLinks=0, status=ImmutableStatusData{statusCode=UNSET, description=}, hasEnded=true}
[otel.javaagent 2025-09-02 22:14:56:074 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'SELECT test' : faa1a2f506b412fc46b6daf1c8bb9bce bc140d85a1f10b67 CLIENT [tracer: io.opentelemetry.jdbc:2.20.0-alpha-SNAPSHOT] AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT * FROM (SELECT n.nspname,c.relname,a.attname,a.atttypid,a.attnotnull OR (t.typtype = ? AND t.typnotnull) AS attnotnull,a.atttypmod,a.attlen,t.typtypmod,row_number() OVER (PARTITION BY a.attrelid ORDER BY a.attnum) AS attnum, nullif(a.attidentity, ?) as attidentity,null as attgenerated,pg_catalog.pg_get_expr(def.adbin, def.adrelid) AS adsrc,dsc.description,t.typbasetype,t.typtype FROM pg_catalog.pg_namespace n JOIN pg_catalog.pg_class c ON (c.relnamespace = n.oid) JOIN pg_catalog.pg_attribute a ON (a.attrelid=c.oid) JOIN pg_catalog.pg_type t ON (a.atttypid = t.oid) LEFT JOIN pg_catalog.pg_attrdef def ON (a.attrelid=def.adrelid AND a.attnum = def.adnum) LEFT JOIN pg_catalog.pg_description dsc ON (c.oid=dsc.objoid AND a.attnum = dsc.objsubid) LEFT JOIN pg_catalog.pg_class dc ON (dc.oid=dsc.classoid AND dc.relname=?) LEFT JOIN pg_catalog.pg_namespace dn ON (dc.relnamespace=dn.oid AND dn.nspname=?) WHERE c.relkind in (?) and a.attnum > ? AND NOT a.attisdropped AND c.relname LIKE ?) c WHERE true ORDER BY nspname,c.relname,attnum }, capacity=128, totalAddedValues=10}
[otel.javaagent 2025-09-02 22:14:56:074 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.javaagent.testing.exporter.OtlpInMemorySpanExporter - Exporting span SpanData{spanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=bc140d85a1f10b67, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, parentSpanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=ebe3a0102b0469cf, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, resource=Resource{schemaUrl=null, attributes={service.instance.id="5b2f150f-21b3-452e-b15d-5af413ba2da8", service.name="unknown_service:java", telemetry.distro.name="opentelemetry-java-instrumentation", telemetry.distro.version="2.20.0-SNAPSHOT", telemetry.sdk.language="java", telemetry.sdk.name="opentelemetry", telemetry.sdk.version="1.53.0"}}, instrumentationScopeInfo=InstrumentationScopeInfo{name=io.opentelemetry.jdbc, version=2.20.0-alpha-SNAPSHOT, schemaUrl=null, attributes={}}, name=SELECT test, kind=CLIENT, startEpochNanos=1756851296053238250, endEpochNanos=1756851296073618584, attributes=AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT * FROM (SELECT n.nspname,c.relname,a.attname,a.atttypid,a.attnotnull OR (t.typtype = ? AND t.typnotnull) AS attnotnull,a.atttypmod,a.attlen,t.typtypmod,row_number() OVER (PARTITION BY a.attrelid ORDER BY a.attnum) AS attnum, nullif(a.attidentity, ?) as attidentity,null as attgenerated,pg_catalog.pg_get_expr(def.adbin, def.adrelid) AS adsrc,dsc.description,t.typbasetype,t.typtype FROM pg_catalog.pg_namespace n JOIN pg_catalog.pg_class c ON (c.relnamespace = n.oid) JOIN pg_catalog.pg_attribute a ON (a.attrelid=c.oid) JOIN pg_catalog.pg_type t ON (a.atttypid = t.oid) LEFT JOIN pg_catalog.pg_attrdef def ON (a.attrelid=def.adrelid AND a.attnum = def.adnum) LEFT JOIN pg_catalog.pg_description dsc ON (c.oid=dsc.objoid AND a.attnum = dsc.objsubid) LEFT JOIN pg_catalog.pg_class dc ON (dc.oid=dsc.classoid AND dc.relname=?) LEFT JOIN pg_catalog.pg_namespace dn ON (dc.relnamespace=dn.oid AND dn.nspname=?) WHERE c.relkind in (?) and a.attnum > ? AND NOT a.attisdropped AND c.relname LIKE ?) c WHERE true ORDER BY nspname,c.relname,attnum }, capacity=128, totalAddedValues=10}, totalAttributeCount=10, events=[], totalRecordedEvents=0, links=[], totalRecordedLinks=0, status=ImmutableStatusData{statusCode=UNSET, description=}, hasEnded=true}
[2025-09-02 22:14:56,082] INFO Checking PostgreSql dialect for type of TABLE "person" (io.confluent.connect.jdbc.dialect.GenericDatabaseDialect)
[otel.javaagent 2025-09-02 22:14:56:086 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'SELECT test' : faa1a2f506b412fc46b6daf1c8bb9bce 501aa90eb31583b9 CLIENT [tracer: io.opentelemetry.jdbc:2.20.0-alpha-SNAPSHOT] AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT NULL AS TABLE_CAT, n.nspname AS TABLE_SCHEM, c.relname AS TABLE_NAME, CASE n.nspname ~ ? OR n.nspname = ? WHEN true THEN CASE WHEN n.nspname = ? OR n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END WHEN n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END END WHEN false THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? then ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE NULL END AS TABLE_TYPE, d.description AS REMARKS, ? as TYPE_CAT, ? as TYPE_SCHEM, ? as TYPE_NAME, ? AS SELF_REFERENCING_COL_NAME, ? AS REF_GENERATION FROM pg_catalog.pg_namespace n, pg_catalog.pg_class c LEFT JOIN pg_catalog.pg_description d ON (c.oid = d.objoid AND d.objsubid = ? and d.classoid = ?::regclass) WHERE c.relnamespace = n.oid AND c.relname LIKE ? AND (false OR ( c.relkind = ? AND n.nspname !~ ? AND n.nspname <> ? ) ) ORDER BY TABLE_TYPE,TABLE_SCHEM,TABLE_NAME }, capacity=128, totalAddedValues=10}
[otel.javaagent 2025-09-02 22:14:56:087 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.javaagent.testing.exporter.OtlpInMemorySpanExporter - Exporting span SpanData{spanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=501aa90eb31583b9, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, parentSpanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=ebe3a0102b0469cf, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, resource=Resource{schemaUrl=null, attributes={service.instance.id="5b2f150f-21b3-452e-b15d-5af413ba2da8", service.name="unknown_service:java", telemetry.distro.name="opentelemetry-java-instrumentation", telemetry.distro.version="2.20.0-SNAPSHOT", telemetry.sdk.language="java", telemetry.sdk.name="opentelemetry", telemetry.sdk.version="1.53.0"}}, instrumentationScopeInfo=InstrumentationScopeInfo{name=io.opentelemetry.jdbc, version=2.20.0-alpha-SNAPSHOT, schemaUrl=null, attributes={}}, name=SELECT test, kind=CLIENT, startEpochNanos=1756851296082378959, endEpochNanos=1756851296086323542, attributes=AttributesMap{data={thread.id=129, db.operation=SELECT, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=SELECT NULL AS TABLE_CAT, n.nspname AS TABLE_SCHEM, c.relname AS TABLE_NAME, CASE n.nspname ~ ? OR n.nspname = ? WHEN true THEN CASE WHEN n.nspname = ? OR n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END WHEN n.nspname = ? THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END END WHEN false THEN CASE c.relkind WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? then ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? WHEN ? THEN ? ELSE NULL END ELSE NULL END AS TABLE_TYPE, d.description AS REMARKS, ? as TYPE_CAT, ? as TYPE_SCHEM, ? as TYPE_NAME, ? AS SELF_REFERENCING_COL_NAME, ? AS REF_GENERATION FROM pg_catalog.pg_namespace n, pg_catalog.pg_class c LEFT JOIN pg_catalog.pg_description d ON (c.oid = d.objoid AND d.objsubid = ? and d.classoid = ?::regclass) WHERE c.relnamespace = n.oid AND c.relname LIKE ? AND (false OR ( c.relkind = ? AND n.nspname !~ ? AND n.nspname <> ? ) ) ORDER BY TABLE_TYPE,TABLE_SCHEM,TABLE_NAME }, capacity=128, totalAddedValues=10}, totalAttributeCount=10, events=[], totalRecordedEvents=0, links=[], totalRecordedLinks=0, status=ImmutableStatusData{statusCode=UNSET, description=}, hasEnded=true}
[2025-09-02 22:14:56,090] INFO Setting metadata for table "person" to Table{name='"person"', type=TABLE columns=[Column{'id', isPrimaryKey=true, allowsNull=false, sqlType=int4}, Column{'name', isPrimaryKey=false, allowsNull=false, sqlType=varchar}]} (io.confluent.connect.jdbc.util.TableDefinitions)
[otel.javaagent 2025-09-02 22:14:56:163 +0000] [task-thread-test-connector-0] DEBUG io.opentelemetry.javaagent.tooling.AgentInstaller$TransformLoggingListener - Transformed org.postgresql.jdbc.PgPreparedStatement -- PluginClassLoader{pluginLocation=file:/usr/share/java/confluentinc-kafka-connect-jdbc/}
[otel.javaagent 2025-09-02 22:14:56:196 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'INSERT test.person' : faa1a2f506b412fc46b6daf1c8bb9bce 999244f75d4a7d29 CLIENT [tracer: io.opentelemetry.jdbc:2.20.0-alpha-SNAPSHOT] AttributesMap{data={thread.id=129, db.operation=INSERT, db.sql.table=person, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=INSERT INTO "person" ("id","name") VALUES (?,?)}, capacity=128, totalAddedValues=11}
[otel.javaagent 2025-09-02 22:14:56:196 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.javaagent.testing.exporter.OtlpInMemorySpanExporter - Exporting span SpanData{spanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=999244f75d4a7d29, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, parentSpanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=ebe3a0102b0469cf, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, resource=Resource{schemaUrl=null, attributes={service.instance.id="5b2f150f-21b3-452e-b15d-5af413ba2da8", service.name="unknown_service:java", telemetry.distro.name="opentelemetry-java-instrumentation", telemetry.distro.version="2.20.0-SNAPSHOT", telemetry.sdk.language="java", telemetry.sdk.name="opentelemetry", telemetry.sdk.version="1.53.0"}}, instrumentationScopeInfo=InstrumentationScopeInfo{name=io.opentelemetry.jdbc, version=2.20.0-alpha-SNAPSHOT, schemaUrl=null, attributes={}}, name=INSERT test.person, kind=CLIENT, startEpochNanos=1756851296189256750, endEpochNanos=1756851296196073625, attributes=AttributesMap{data={thread.id=129, db.operation=INSERT, db.sql.table=person, db.name=test, server.port=5432, thread.name=task-thread-test-connector-0, db.user=postgres, db.connection_string=postgresql://postgres:5432, server.address=postgres, db.system=postgresql, db.statement=INSERT INTO "person" ("id","name") VALUES (?,?)}, capacity=128, totalAddedValues=11}, totalAttributeCount=11, events=[], totalRecordedEvents=0, links=[], totalRecordedLinks=0, status=ImmutableStatusData{statusCode=UNSET, description=}, hasEnded=true}
[otel.javaagent 2025-09-02 22:14:56:201 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.exporter.logging.LoggingSpanExporter - 'KafkaConnect.put' : faa1a2f506b412fc46b6daf1c8bb9bce ebe3a0102b0469cf INTERNAL [tracer: io.opentelemetry.kafka-connect-2.6:2.20.0-alpha-SNAPSHOT] AttributesMap{data={thread.id=129, thread.name=task-thread-test-connector-0}, capacity=128, totalAddedValues=2}
[otel.javaagent 2025-09-02 22:14:56:202 +0000] [task-thread-test-connector-0] INFO io.opentelemetry.javaagent.testing.exporter.OtlpInMemorySpanExporter - Exporting span SpanData{spanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=ebe3a0102b0469cf, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, parentSpanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=e37a2c4da90a8684, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=true, valid=true}, resource=Resource{schemaUrl=null, attributes={service.instance.id="5b2f150f-21b3-452e-b15d-5af413ba2da8", service.name="unknown_service:java", telemetry.distro.name="opentelemetry-java-instrumentation", telemetry.distro.version="2.20.0-SNAPSHOT", telemetry.sdk.language="java", telemetry.sdk.name="opentelemetry", telemetry.sdk.version="1.53.0"}}, instrumentationScopeInfo=InstrumentationScopeInfo{name=io.opentelemetry.kafka-connect-2.6, version=2.20.0-alpha-SNAPSHOT, schemaUrl=null, attributes={}}, name=KafkaConnect.put, kind=INTERNAL, startEpochNanos=1756851295420017000, endEpochNanos=1756851296201312542, attributes=AttributesMap{data={thread.id=129, thread.name=task-thread-test-connector-0}, capacity=128, totalAddedValues=2}, totalAttributeCount=2, events=[], totalRecordedEvents=0, links=[ImmutableLinkData{spanContext=ImmutableSpanContext{traceId=faa1a2f506b412fc46b6daf1c8bb9bce, spanId=e37a2c4da90a8684, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=true, valid=true}, attributes={}, totalAttributeCount=0}], totalRecordedLinks=1, status=ImmutableStatusData{statusCode=UNSET, description=}, hasEnded=true}
```

I see that, we have perfect trace propagation:
Producer span: e37a2c4da90a8684 (remote=true, the original message)
Kafka Connect span: ebe3a0102b0469cf (parent: e37a2c4da90a8684, with span link)
JDBC spans: All have parent ebe3a0102b0469cf, including the crucial INSERT test.person operation

3. Regarding the span kind: I felt Kafka Connect acts more like an internal processing/transformation layer rather than a direct consumer client. It receives messages, processes them (potentially transforming data), and forwards them to sinks. This seems more aligned with internal processing than direct message consumption. Tbh I started developing this as a CONSUMER and shifted it to INTERNAL midway.

Should I change this to CONSUMER as I am conflicted? I'm open to feedback on this decision.

laurit commented:
> 3\. Regarding the span kind: I felt Kafka Connect acts more like an internal processing/transformation layer rather than a direct consumer client. It receives messages, processes them (potentially transforming data), and forwards them to sinks. This seems more aligned with internal processing than direct message consumption. Tbh I started developing this as a CONSUMER and shifted it to INTERNAL midway.
> 
> Should I change this to CONSUMER as I am conflicted? I'm open to feedback on this decision.

Doesn't it read from the topic like a regular consumer? If you believe that `INTERNAL` span is the correct choice then keep in mind that the convention in this project is not to emit telemetry that doesn't have semantic conventions in default configuration. This means that we usually instrumentations that only emit internal spans need to be enabled with a flag e.g https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/3998e96c75dd7b1de272693815859c98b1edc6fa/instrumentation/mybatis-3.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/mybatis/v3_2/MyBatisInstrumentationModule.java#L30

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/KafkaConnectTask.java:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  public static String getSpanName(KafkaConnectTask task) {
+    return "KafkaConnect.put";
```
It would be better to use ` MessagingSpanNameExtractor.create(getter, operation)` for the span name since `KafkaConnect.put` doesn't conform to messaging span name conventions https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/messaging/messaging-spans.md#span-name

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/SinkTaskInstrumentation.java:
```diff
@@ -0,0 +1,83 @@
+      Context parentContext = Java8BytecodeBridge.currentContext();
+
+      // Extract context from first record if available
+      if (!records.isEmpty()) {
+        SinkRecord firstRecord = records.iterator().next();
+        Context extractedContext =
+            KafkaConnectSingletons.propagator()
+                .extract(
+                    parentContext, firstRecord, KafkaConnectSingletons.sinkRecordHeaderGetter());
+        parentContext = extractedContext;
+      }
```
Why the first record? If there are going to be multiple records from different traces then maybe should use `Context.current()` as the parent and rely on the span links for connecting traces?

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/SinkTaskInstrumentation.java:
```diff
@@ -0,0 +1,83 @@
+      }
+
+      task = new KafkaConnectTask(records);
+      if (!KafkaConnectSingletons.instrumenter().shouldStart(parentContext, task)) {
```
usually we static import the `instrumenter` method

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,41 @@
+  testImplementation("org.apache.kafka:connect-runtime:3.6.1")
+  testImplementation(project(":instrumentation:kafka:kafka-clients:kafka-clients-0.11:testing"))
+
+  implementation("org.testcontainers:postgresql:1.21.3") // For PostgreSQLContainer
```
since this module only has tests it is strange to see a mix of `implementation` and `testImplementation` dependencies

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,41 @@
+
+  // Testcontainers dependencies for integration testing
+  testImplementation("io.strimzi:strimzi-test-container:0.111.0")
+  implementation("org.testcontainers:junit-jupiter")
```
is this needed?

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,41 @@
+  implementation(project(":instrumentation:kafka:kafka-clients:kafka-clients-0.11:javaagent"))
+
+  testImplementation("org.apache.kafka:connect-runtime:3.6.1")
+  testImplementation(project(":instrumentation:kafka:kafka-clients:kafka-clients-0.11:testing"))
```
not needed?

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,611 @@
+    String kafkaConnectSpanId = null;
+    int spanCount = 0;
+
+    for (JsonNode trace : tracesNode) {
```
Can't really tell what this is asserting, I think we need to come up with a better solution for this.

vasantteja replied:
I agree with this. I’ve redesigned the entire testing approach to verify whether end-to-end tracing is fully implemented.

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,41 @@
+  implementation("com.fasterxml.jackson.core:jackson-databind")
+  testImplementation("com.squareup.okhttp3:okhttp:4.12.0")
+
+  compileOnly("com.google.auto.value:auto-value-annotations")
+  annotationProcessor("com.google.auto.value:auto-value")
```
These are not needed. I removed them.

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,41 @@
+  implementation(project(":instrumentation:kafka:kafka-connect-2.6:javaagent"))
+  implementation(project(":instrumentation:kafka:kafka-clients:kafka-clients-0.11:javaagent"))
+
+  testImplementation("org.apache.kafka:connect-runtime:3.6.1")
```
Yep, I replaced it with kafka-clients.

vasantteja commented on javaagent/build.gradle.kts:
```diff
@@ -93,6 +93,7 @@ dependencies {
   baseJavaagentLibs(project(":instrumentation:opentelemetry-api:opentelemetry-api-1.50:javaagent"))
   baseJavaagentLibs(project(":instrumentation:opentelemetry-api:opentelemetry-api-1.52:javaagent"))
   baseJavaagentLibs(project(":instrumentation:opentelemetry-instrumentation-api:javaagent"))
+  baseJavaagentLibs(project(":instrumentation:kafka:kafka-connect-2.6:javaagent"))
```
Yeah I added the agent here as I am attaching it to the kafka-connect testcontainer. With out this shadowJar my instrumentation is not getting attached to the kafka-connect testcontainer.

vasantteja replied:
I removed after pivoting to the strategy used in smoke tests. Please ignore my previous comment.

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,41 @@
+}
+
+dependencies {
+  api(project(":testing-common"))
```
Gotcha, I removed it.

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,41 @@
+plugins {
+  id("otel.java-conventions")
+  id("otel.javaagent-testing")
```
Thanks this makes a lot of sense. I pivoted to the strategy used in smoke tests and it's attaching the instrumentation as a shadow jar to the containers and I can see spans for all the operations.

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,53 @@
+  )
+
+  // Pass backend endpoint as system property (will be set dynamically in test)
+  systemProperty("otel.exporter.otlp.traces.endpoint", System.getProperty("otel.exporter.otlp.traces.endpoint", "http://localhost:4318/v1/traces"))
```
this won't work

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/build.gradle.kts:
```diff
@@ -0,0 +1,53 @@
+  // Make agent jar path available to tests
+  systemProperty("io.opentelemetry.smoketest.agent.shadowJar.path", agentShadowJar.get().archiveFile.get().toString())
+  // Configure test JVM with agent for end-to-end tracing
+  jvmArgs(
+    "-javaagent:${agentShadowJar.get().archiveFile.get()}",
+    "-Dotel.traces.exporter=otlp",
+    "-Dotel.metrics.exporter=none",
+    "-Dotel.logs.exporter=none",
+    "-Dotel.service.name=kafka-connect-test-producer"
+  )
```
is this really useful

vasantteja replied:
No this is not useful. I removed them.

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,728 @@
+
+@Testcontainers
+// Suppressing warnings for test dependencies and deprecated Testcontainers API
+@SuppressWarnings({"rawtypes", "unchecked", "deprecation", "unused"})
```
try suppressing less warning, you seem to have a bunch of unused fields and variables

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,728 @@
+
+  // Docker network / containers
+  private static Network network;
+  private static FixedHostPortGenericContainer<?> kafka;
```
Is using the deprecated `FixedHostPortGenericContainer` necessary? In other tests we use `KafkaContainer`, for example see https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/kafka/kafka-clients/kafka-clients-0.11/testing/src/main/java/io/opentelemetry/instrumentation/kafkaclients/common/v0_11/internal/KafkaClientBaseTest.java Perhaps that would help?

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,728 @@
+  public void testKafkaConnectMongoSinkTaskInstrumentation()
+      throws IOException, InterruptedException {
+    // Create unique topic name
+    String uniqueTopicName = TOPIC_NAME + "-" + System.currentTimeMillis();
```
It is best not to do this but rather use the `TOPIC_NAME `. Messaging semantic conventions require that span names have low cardinality https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#span-name. Here you make the topic name look like it was a temporary name so whoever is looking at it will think that there must be a bug in the instrumentation because such name shouldn't be captured in span name.

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/PostgresKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,759 @@
+@Testcontainers
+// Suppressing warnings for test dependencies and deprecated Testcontainers API
+@SuppressWarnings({"deprecation", "unused"})
+class PostgresKafkaConnectSinkTaskTest {
```
would it make sense to place common code between mongo and postgres tests into a common superclass?

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,728 @@
+
+    // Perform distributed tracing assertions
+    // Assertion 1: Verify all spans and links are not null
+    assertThat(tracingData.kafkaConnectConsumerSpan)
```
@laurit I believe after merging this [PR](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/14731) I can use SmokeTestInstrumentationExtension to do complete end to end assertion on traceId and spanId similar to other kafka instrumentations.

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/KafkaConnectTask.java:
```diff
@@ -0,0 +1,31 @@
+
+  /**
+   * Returns the first record in the batch, used for extracting destination information. Kafka
+   * Connect processes records in batches, but all records in a batch typically come from the same
```
@laurit These records can come from multiple topics. When Kafka Connect processes records from multiple topics in a single batch, I've updated the span naming convention to reflect all topics included in the batch. For example, the new span name will be something like [topic-1, topic-2, topic-3] process instead of just using a single topic name. This provides better visibility into which topics are being processed together in a single batch operation.

vasantteja commented:
@laurit Can you please take one more look? I refactored and made the following changes:

1. Changed the span naming algorithm to include all the topics instead of the first record topic name.
2. Remove the parent child relationship between producer and kafka connect spans.
3. Modularized and rewrote the tests with new `SmokeTestInstrumentationExtension`

Let me know what do you think.

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/KafkaConnectSingletons.java:
```diff
@@ -0,0 +1,60 @@
+    return PROPAGATOR;
+  }
+
+  public static TextMapGetter<SinkRecord> sinkRecordHeaderGetter() {
+    return SINK_RECORD_HEADER_GETTER;
+  }
```
this is also not used

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/SinkRecordHeadersGetter.java:
```diff
@@ -0,0 +1,47 @@
+  @Override
+  public Iterable<String> keys(SinkRecord record) {
+    if (record.headers() == null) {
+      return java.util.Collections.emptyList();
```
could import `Collections`

laurit commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,352 @@
+  @Override
+  protected void stopDatabaseContainer() {
+    if (mongoDB != null) {
+      try {
+        mongoDB.stop();
+      } catch (RuntimeException e) {
+        logger.error("Error stopping MongoDB: " + e.getMessage());
+      }
```
Usually we just let the errors propagate. No need to catch anything unless you expect it to fail and need to catch the exception to avoid it breaking stuff.

vasantteja replied:
I removed the catch blocks but left them around handful of places as compiler was complaining about lack of error handling.

vasantteja commented:
@laurit Can you please take another look? Kafka connect's [WorkerSinkTask](https://github.com/apache/kafka/blob/611f4128b4a38795be9a6d928c60e55a41b93f3d/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L328) is pulling records from multiple topics instead of one single topic. I am returning the topic name if all the records are from single topic otherwise I am returning a null. Here is the `SpanData`

`{ "traceId" : "835d2dc1437d7f420a28adfab89347bb", "spanId" : "d9db28e45ce804f7", "startEpochNanos" : 1759528513964221000, "endEpochNanos" : 1759528514783972376, "kind" : "CONSUMER", "name" : "unknown process", "attributes" : { "messaging.system" : "kafka", "messaging.batch.message_count" : 3, "thread.name" : "task-thread-test-postgres-connector-multi-0", "thread.id" : 134, "messaging.operation" : "process" }, "parentSpanId" : "0000000000000000" }`

Please let me know if you need additional changes.

laurit commented on instrumentation/kafka/kafka-connect-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaconnect/v2_6/KafkaConnectTask.java:
```diff
@@ -0,0 +1,46 @@
+    if (topics.size() == 1) {
+      return topics.iterator().next();
+    }
+    return topics.stream().collect(Collectors.joining(",", "[", "]"));
```
thanks for looking into it

jaydeluca commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,354 @@
+                span ->
+                    span.hasName("kafka-connect-status process")
+                        .hasKind(SpanKind.CONSUMER)
+                        .hasParent(trace.getSpan(0)));
```
i noticed you aren't asserting any attributes, is that intentional?

jaydeluca commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -0,0 +1,354 @@
+  }
+
+  @Test
+  public void testKafkaConnectMongoSinkTaskMultiTopicInstrumentation() throws Exception {
```
this note applies to all test methods, methods don't need to be public

```suggestion
  void testKafkaConnectMongoSinkTaskMultiTopicInstrumentation() throws Exception {
```

vasantteja commented on instrumentation/kafka/kafka-connect-2.6/javaagent/build.gradle.kts:
```diff
@@ -0,0 +1,19 @@
+  pass {
+    group.set("org.apache.kafka")
+    module.set("connect-api")
+    versions.set("[2.6.0,)")
```
what does assertInverse do?
